File: archive/task26/task26.py
def remove_intersections(intervals):
    result = []

    s = set(range(intervals[0][0], intervals[0][1] + 1))

    i = 1
    while i < len(intervals):
        new_interval = set(range(intervals[i][0], intervals[i][1] + 1))
        if not s.isdisjoint(new_interval):
            s = s.union(new_interval)
            i += 1
        else:
            l = list(s)
            s = set(range(intervals[i][0], intervals[i][1] + 1))
            result.append((l[0], l[-1]))

    if s != {}:
        l = list(s)
        result.append((l[0], l[-1]))

    return result


def test1():
    intervals = [(1, 5), (2, 3), (4, 6), (7, 8), (8, 10), (12, 15)]
    expected = [(1, 6), (7, 10), (12, 15)]

    print(remove_intersections(intervals) == expected)


def test2():
    intervals = [(9, 10), (10, 12), (11, 13), (14, 18), (15, 18)]
    expected = [(9, 13), (14, 18)]

    print(remove_intersections(intervals) == expected)


def main():
    test1()
    test2()


if __name__ == '__main__':
    main()

File: archive/task26/task26.cpp
#include <iostream>
#include <vector>
#include <stack>

typedef std::pair<int, int> Interval;

std::vector<Interval> non_intersecting(const std::vector<Interval> &intervals)
{
    std::vector<Interval> result;

    int len_intervals = intervals.size();

    std::stack<Interval> interval_stack;
    interval_stack.push(intervals[len_intervals - 1]);

    for (int i = len_intervals - 2; i >= 0; --i)
    {
        Interval &top = interval_stack.top();

        if (intervals[i].second < top.first)
        {
            interval_stack.push(intervals[i]);
        }
        else
        {
            while (!interval_stack.empty() && intervals[i].second >= interval_stack.top().first)
            {
                top = interval_stack.top();
                interval_stack.pop();
            }

            interval_stack.push({
                std::min(intervals[i].first, top.first),
                std::max(intervals[i].second, top.second)
            });
        }
    }

    while (!interval_stack.empty())
    {
        result.push_back(interval_stack.top());
        interval_stack.pop();
    }

    return result;
}

void test1()
{
    std::vector<Interval> intervals{{1, 5}, {2, 3}, {4, 6}, {7, 8}, {8, 10}, {12, 15}};

    for (const Interval &i : non_intersecting(intervals))
    {
        std::cout << '{' << i.first << ", " << i.second << "} ";
    }
    std::cout << '\n';
}

void test2()
{
    std::vector<Interval> intervals{{9, 10}, {10, 12}, {11, 13}, {14, 18}, {15, 18}};

    for (const Interval &i : non_intersecting(intervals))
    {
        std::cout << '{' << i.first << ", " << i.second << "} ";
    }
    std::cout << '\n';
}

int main(int argc, char const *argv[])
{
    test1(); // {1, 6} {7, 10} {12, 15}
    test2(); // {9, 13} {14, 18}

    return 0;
}
